librad/src/core.rs

```diff
@@ -0,0 +1,276 @@
+
+        url.set_path(
+            Path::new(&multibase::encode(Base::Base32Z, self.urn.id))
+                .join(self.urn.path)
+                .to_str()
+                .unwrap(),
+        );
+
+        url
+    }
+}
+
+/// Placeholder for a version in a history
+type Version<'a> = &'a [u8];
+
+/// Placeholder for the data passed to `Verifier::verify`
+pub struct Rev<'a> {
+    pub version: &'a Version<'a>,
+    pub payload: &'a [u8],
+}
+
+/// A verification function for an identity history as named by a `RadUrn`.
+///
+/// The supplied iterator traverses the history in reverse order, i.e.
+/// oldest-first.
+///
+/// In order to satisfy the verification requirements, `Verifier::verify` may
+/// call `Core::fetch` recursively.
+#[async_trait]
+pub trait Verifier {
+    type Error;
+
+    async fn verify<'a>(
+        history: Box<dyn Iterator<Item = Rev<'a>>>,
+    ) -> Result<&'a Version<'a>, Self::Error>;
+}
+
+pub struct Refsig<'a> {
+    pub refs: HashMap<&'a Path, &'a [u8]>,
+    pub signature: Signature,
+}
+
+pub enum BrowseError {
+    NotConnected,
+}
+
+#[async_trait]
+pub trait Browse {
+    type Stream: Stream<Item = RadUrn>;
+
+    /// Given a known peer, ask it to enumerate all [`RadUrn`]s it knows about.
+    ///
+    /// This is an online query: if no connection to the peer exists, or could
+    /// be established, an error is returned.
+    async fn browse(&self, peer: &PeerId) -> Result<Self::Stream, BrowseError>;
+
+    /// Peek at the most recent `Rev` of the `RadUrn`.
+    ///
+    /// See also [`Fetch::fetch`].
+    async fn peek<'a>(&self, peer: &PeerId, urn: &RadUrn) -> Result<Rev<'a>, BrowseError>;
+}
+
+pub struct Have {
+    pub entity: RadUrn,
+    pub head: Multihash,
+}
+
+#[async_trait]
+pub trait Gossip {
+    type QueryStream: Stream<Item = PeerId>;
+
+    /// Announce an update to a local repository to the network.
+    ///
+    /// Nb.: the update refers to the "owned" branches of the repository, i.e.
+    /// `refs/heads`. Precondition: `rad/refsigs` has been updated.
+    async fn announce(&self, have: Have);
+
+    /// Find peers on the network which provide [`RadUrn`].
+    ///
+    /// If `head` is given, restrict to peers who can provide this revision or
+    /// later.
+    ///
+    /// The query may be answered from the local cached view of the network. The
+    /// caller controls how many peers to fetch, and for how long, by either
+    /// continuing to poll the `Stream` or dropping it.
+    fn query(&self, urn: &RadUrn, head: &Multihash) -> Self::QueryStream;
+}
+
+#[non_exhaustive]
+pub enum FetchError<V> {
+    Verification(V),
+    NoSuchBranch(PathBuf),
+    Io(io::Error),
+    // ...
+}
+
+#[non_exhaustive]
+pub enum ShallowFetchError {
+    NoSuchBranch(PathBuf),
+    Io(io::Error),
+    // ...
+}
+
+#[async_trait]
+pub trait Fetch {
+    /// Iterator over the commit graph starting at the head of the branch
+    /// specified by a [`RadUrl`].
+    type Revwalk: Iterator;
```

Comment:
I wonder if a [`DoubleEndedIterator`](https://doc.rust-lang.org/std/iter/trait.DoubleEndedIterator.html) might be a better shout here. Since we want to walk from the back.

Reply:
Well, but the verifier fun walks it in only one direction. Also, neither `git2::Revwalk` nor `surf::History` implement this _hint hint_

Reply:
Soooo, you want me to create a ticket to add that? :stuck_out_tongue_closed_eyes: